### PR TITLE
MSVC Build fixes

### PR DESCRIPTION
**Pull request by izzyonstage**
Unsure how to separate these commits into different pull requests (sorry).

These changes are for everything I had to do in order to build QtWebKit and be able to use it with PhantomJS.

Fixes from Issues: #579 and #578

### CODE REVIEW

**Comment by izzyonstage** [comment]
> Get library name from cmake, i.e. parse ${LIBXML2_LIBRARIES} etc

I was not sure if that would work. Currently I am having to specify that as `libxml2_a.lib` which would be wrong for this usage i believe?

File: Source/WebCore/PlatformQt.cmake [hunk 1]
```diff
@@ -437,6 +437,10 @@ if (WIN32)
         add_definitions(-DSTATICALLY_LINKED_WITH_WTF -DSTATICALLY_LINKED_WITH_JavaScriptCore)
     endif ()
 
+    list(APPEND WebCore_LIBRARIES
```
Comment (annulen) [inline Source/WebCore/PlatformQt.cmake]: IIRC this should not be needed anymore after @Vitallium updated  qtwebkit-win-libs.zip

**Comment by annulen** [comment]
CMake variables like LIBXML2_LIBRARIES,  SQLITE_LIBRARIES etc. contain full paths fo library files. You should extract library names from them using built-in cmake function `get_filename_component` (see docs). You should handle cases when there are several items in each variables (they are supposed to be lists), or when they are empty, or contain something that `get_filename_component` fails to parse

File: Source/WebCore/PlatformQt.cmake (hunk 1 above)
Comment (izzyonstage) [inline Source/WebCore/PlatformQt.cmake]: Will double check with my build and remove if it works without this.

File: Source/cmake/OptionsQt.cmake [hunk 2]
```diff
@@ -282,6 +283,9 @@ if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
     endif ()
 elseif (MSVC AND MSVC_VERSION LESS 1900)
     message(FATAL_ERROR "MSVC 2015 is required to build QtWebKit, use a newer MSVC version")
+elseif (MSVC)
+    # Required due to changes in MSVC 2015 CRT functions now being inline or moved to different headers
+    set(CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES} legacy_stdio_definitions.lib)
```
Comment (Vitallium) [inline Source/cmake/OptionsQt.cmake]: Could you please elaborate more on this?

File: Source/WebKit/PlatformQt.cmake [hunk 3]
```diff
@@ -758,9 +755,10 @@ install(TARGETS WebKitWidgets EXPORT Qt5WebKitWidgetsTargets
         DESTINATION "${LIB_INSTALL_DIR}"
         RUNTIME DESTINATION "${BIN_INSTALL_DIR}"
 )
-if (MSVC)
-    install(FILES $<TARGET_PDB_FILE:WebKitWidgets> DESTINATION "${BIN_INSTALL_DIR}" OPTIONAL)
-endif ()
+# temporarily removed as only required if building with PDB output files
+#if (MSVC)
```
Comment (annulen) [inline Source/WebKit/PlatformQt.cmake]: I'll push fix for this after it passes CI

Comment (Vitallium) [inline Source/WebKit/PlatformQt.cmake]: Nice! Thanks.

File: Source/cmake/OptionsQt.cmake (hunk 2 above)
Comment (izzyonstage) [inline Source/cmake/OptionsQt.cmake]: [https://eldeeb.net/wrdprs/?p=386](https://eldeeb.net/wrdprs/?p=386)
I was getting some of the same build errors as in the above link.
After lots of googling I found that the CRT headers from MSVC2015 have changed in order to conform to the standards.
[https://msdn.microsoft.com/en-us/library/bb531344.aspx](https://msdn.microsoft.com/en-us/library/bb531344.aspx)
They have suggested adding this legacy lib which will expose the required definitions as they were seen before, thus fixing the missing definitions.
The only place I found this making any difference was by adding it to the `CMAKE_REQUIRED_LIBRARIES` variable as then it gets included everywhere.

I agree this is really only a quick fix for now, but this was the quickest way to get my build to work.

Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: @izzyonstage I think we don't use NODEFAULTLIB, do we?

Comment (izzyonstage) [inline Source/cmake/OptionsQt.cmake]: something must be somewhere, as without this I am unable to build using MSVC2015.

File: Source/WebCore/PlatformQt.cmake (hunk 1 above)
Comment (izzyonstage) [inline Source/WebCore/PlatformQt.cmake]: Just wondering....

> should not be needed anymore after @Vitallium updated qtwebkit-win-libs.zip

That may not affect me as I am building all the dependencies myself from scratch anyway so I don't use that zip.
I have removed that change in my fork and will see what happens in my CI.

File: Source/WebKit/PlatformQt.cmake [hunk 4]
```diff
@@ -446,17 +446,27 @@ if (ENABLE_PRINT_SUPPORT)
...
 if (QT_STATIC_BUILD)
-    if (MSVC)
-        set(LIB_PREFIX "lib")
-    endif ()
     set(WEBKITWIDGETS_PKGCONGIG_DEPS "${WEBKITWIDGETS_PKGCONGIG_DEPS} Qt5PrintSupport")
     set(WEBKITWIDGETS_PRI_DEPS "${WEBKITWIDGETS_PRI_DEPS} printsupport")
-    set(EXTRA_LIBS_NAMES WebCore JavaScriptCore WTF xml2)
+    set(EXTRA_LIBS_NAMES WebCore JavaScriptCore WTF)
+    if (MSVC)
+        foreach (LIB_FILENAME ${LIBXML2_LIBRARIES})
```
Comment (annulen) [inline Source/WebKit/PlatformQt.cmake]: Please extract repeating code into macro

File: Source/WebKit/PlatformQt.cmake [hunk 5]
```diff
@@ -446,17 +446,27 @@ if (ENABLE_PRINT_SUPPORT)
...
+            get_filename_component(LIB_NAME ${LIB_FILENAME} NAME_WE)
+		    list(APPEND EXTRA_LIBS_NAMES ${LIB_NAME})
+        endforeach ()
+        foreach (LIB_FILENAME ${SQLITE_LIBRARIES})
+            get_filename_component(LIB_NAME ${LIB_FILENAME} NAME_WE)
+		    list(APPEND EXTRA_LIBS_NAMES ${LIB_NAME})
+        endforeach ()
+        list(APPEND EXTRA_LIBS_NAMES zlib)
```
Comment (annulen) [inline Source/WebKit/PlatformQt.cmake]: ZLIB_LIBRARIES

File: Source/JavaScriptCore/CMakeLists.txt [hunk 6]
```diff
@@ -985,7 +985,7 @@ target_link_libraries(LLIntOffsetsExtractor WTF)
 # LLIntOffsetsExtractor matches, no output is generated. To make this target consistent and avoid
 # running this command for every build, we artificially update LLIntAssembly.h's mtime (using touch)
 # after every asm.rb run.
-if (MSVC)
+if (MSVC AND ENABLE_JIT)
```
Comment (annulen) [inline Source/JavaScriptCore/CMakeLists.txt]: unused?

File: Source/WebCore/PlatformQt.cmake (hunk 1 above)
Comment (izzyonstage) [inline Source/WebCore/PlatformQt.cmake]: this does seem to now work without this change. I have removed in my fork

File: Source/cmake/OptionsQt.cmake (hunk 2 above)
Comment (izzyonstage) [inline Source/cmake/OptionsQt.cmake]: After removing this it does seem that it is not required now. must have been a broken build of one of the dependencies I had. Removed this in my fork

File: Source/JavaScriptCore/CMakeLists.txt (hunk 6 above)
Comment (izzyonstage) [inline Source/JavaScriptCore/CMakeLists.txt]: this may have been a change from a previous build before I started using jom

Comment (izzyonstage) [inline Source/JavaScriptCore/CMakeLists.txt]: however when not using jom this was a big issue, although even with this change it was still broken further in the build pipeline

Comment (annulen) [inline Source/JavaScriptCore/CMakeLists.txt]: JIT has nothing to do with jom except first letter

File: Source/WebKit/PlatformQt.cmake (hunk 4 above)
Comment (izzyonstage) [inline Source/WebKit/PlatformQt.cmake]: I have done exactly that in my local repo, just have not yet pushed the changes.
Having an issue with DumpRenderTree not finding zlib (not sure why it needs it though).

EDIT:
I believe the way I built libxml2 is the reason for the zlib dependency. just running new build now to confirm.

File: Source/WebKit/PlatformQt.cmake (hunk 5 above)
Comment (izzyonstage) [inline Source/WebKit/PlatformQt.cmake]: I have added that change also in my local repo. will push this change shortly

File: Source/JavaScriptCore/CMakeLists.txt (hunk 6 above)
Comment (izzyonstage) [inline Source/JavaScriptCore/CMakeLists.txt]: originally I was using nmake directly and not jom

**Comment by izzyonstage** [comment]
Looks like I am stuck with issues for DumpRenderTree.

Where should I add the zlib library for the linker to work?

**Comment by izzyonstage** [comment]
If it is OK I am going to close this pull request.
Will be recreating my fork as it is completely broken on my local machine now.
Hoping to get latest and greatest commits and apply my changes in different branches to make specific pull requests for any remaining changes I make.

**Comment by annulen** [comment]
That's fine. In future you may want to keep your changes in a branch and rebase it when upstream updates. Use `rebase -i` to fix your previous patches

**Pull request closed without merging**